game-core/src/main/java/games/strategy/triplea/ai/pro/util/ProMoveUtils.java

```diff
@@ -178,6 +230,7 @@ public static void calculateAmphibRoutes(
 
     final GameData data = ProData.getData();
 
+    final HashMap<Territory, ArrayList<Move>> movesMap = new HashMap<>();
```

Comment:
I'd appreciate if you could try to use interface types wherever possible, I.e.
`Map` instead of `HashMap` or `List` instead of `ArrayList`.
This way we could easily swap out the implementation at a later time if it no longer servers our needs

Reply:
This applies to all other parts of your PR as well btw

Reply:
Calling methods on interfaces is slower than on concrete tables. The latter is done through a vtable (so just get the pointer at the given offset), while the former requires scanning through an itable, see:
https://lukasatkinson.de/2018/interface-dispatch/#java-itables

So in local code (i.e. not in parameters and return values), I do prefer using concrete types for that reason. For actual APIs, keeping things as interfaces makes sense - since we don't want to make restrictions on callers. But when it's all local (within a function or in a private helper class), why make the code slower for very minimal benefit?

Reply:
Interesting article you linked there 👍

However the article also notes that in practice the JIT Compiler optimizes away some of that slower lookup, so it likely won't be noticeable at all.

If you really insist on having concrete types for local variables, I'd kindly ask you to use the java 11 `var` keyword to keep the code shorter, but I personally dislike having concrete implementations even for helper methods. We ran into some issues in the past that make it sheer impossible to change the implementation type, because it is being passed so many layers deep and I really want to avoid that.

Reply:
True that JIT should help. I like the var suggestion for locals. Will go with that.

Reply:
I'd back going with the more generic type:
- it's a pseudo-java convention, not just in this project, but has been going for some time. The 'least-surprise' principle means to try and follow expectations of other programmers
- it encourages to not depend concrete types unnecessarily, good point about locals, though even then there is something to be said about having just the type or API you need.
- we should be careful of micro-optimizations. The runtime difference of tracing a few a pointers vs one is pretty certainly order nano-second. We have cases where game-data fetches are retrieved over network for clients, which is 6 to 9 orders of magnitude more significant. To that point, I really recommend the code-horror article [the infinite space between words](https://blog.codinghorror.com/the-infinite-space-between-words/), it illustrates that point well. To that extent, the JVM is quite good about optimizations that at a local level, you don't need to worry about it so much.
- last, for the most part the rest of the codebase uses as generic of types as possible, it's worthwhile to keep it consistent

> I like the var suggestion for locals 

Let's be careful to use this well. So far, AFAIK, var is best used when you have a local where the name and type match, eg: `Map< > map`, or something like: `Image image`. Names that are identical to the type can be a code smell. Regardless, using `var` can be dangerous if the type information is not clear.

Reply:
I agree that var is not good when type info is not clear - e.g. if you're assigning the return value of some method. But if we're just creating a local collection - i.e. = new HashMap<...>(); then I think that's not a concern. Do you agree?

i.e. the suggestion is for a line like:
final HashMap<Territory, ArrayList<Move>> movesMap = new HashMap<>();

To become:

final var movesMap = new HashMap<Territory, ArrayList<Move>>();

So we're not losing any type info there.

Reply:
Sounds good, for reference,  [Effective Java Item 64](https://kea.nu/files/textbooks/new/Effective%20Java%20%282017%2C%20Addison-Wesley%29.pdf)  "Item 64:  Refer to objects by their interfaces"

Reply:
OK, I'll go with var for local collections that are created with new and with interface types otherwise. I think this should be a good compromise. Done.

Reply:
Uh. I think adopting broad use of `var` is probably something that is a broader discussion. I don't particularly like it and generally I haven't seen it used much in Java projects. It would be better to do a refactoring just focusing on `var` rather than introducing it here.

Reply:
Agree broader discussion, couple quick mentions:
- We've already 58 uses

```
$ find . -name "*.java" | xargs cat | grep -c "var " 
58
```
- not a lot of projects are java 10+
- usage needs to be strategic when type information is obvious, eg: `Image image = new Image()` to `var image = new Image()`

Effective Java does not yet have a recommendation on var, I'm pretty sure it would be "use var judiciously" with careful notes about typing, similar to [(P203) - Item 45](https://kea.nu/files/textbooks/new/Effective%20Java%20%282017%2C%20Addison-Wesley%29.pdf) "Use streams judiciously"

> Overusing streams makes programs hard to read and maintain

>  In  the  absence  of  explicit  types,  careful  naming  of  lambdaparameters is essential to the readability of stream pipelines.